buoy_gazebo/src/buoy_utils/StopwatchSimTime.hpp

```diff
@@ -0,0 +1,182 @@
+// Copyright 2022 Open Source Robotics Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BUOY_UTILS__STOPWATCHSIMTIME_HPP_
+#define BUOY_UTILS__STOPWATCHSIMTIME_HPP_
+
+#include <rclcpp/clock.hpp>
+
+#include <chrono>
+#include <memory>
+
+
+namespace buoy_utils
+{
+// Forward declarations.
+class StopwatchSimTimePrivate;
+
+//
+/// \class StopwatchSimTime StopwatchSimTime.hh buoy_utils/StopwatchSimTime.hh
+/// \brief The stopwatch keeps track of time spent in the run state,
+/// accessed through ElapsedRunTime(), and time spent in the stop state,
+/// accessed through ElapsedStopTime(). Elapsed run time starts accumulating
+/// after the first call to Start(). Elapsed stop time starts
+/// accumulation after Start() has been called followed by Stop(). The
+/// stopwatch can be reset with the Reset() function.
+///
+/// # Example usage
+///
+/// ```{.cpp}
+/// buoy_utils::StopwatchSimTime watch;
+/// watch.Start();
+///
+/// // do something...
+///
+/// std::cout << "Elapsed time is "
+/// << std::chrono::duration_cast<std::chrono::milliseconds>(
+///   timeSys.ElapsedRunTime()).count() << " ms\n";
+/// watch.Stop();
+/// ```
+class StopwatchSimTime
+{
+  /// \brief Constructor.
+
+public:
+  StopwatchSimTime();
+
+  /// \brief Copy constructor
+  /// \param[in] _watch The stop watch to copy.
+
+public:
+  StopwatchSimTime(const StopwatchSimTime & _watch);
+
+  /// \brief Move constructor
+  /// \param[in] _watch The stop watch to move.
+
+public:
+  StopwatchSimTime(StopwatchSimTime && _watch) noexcept;
+
+  /// \brief Destructor.
+
+public:
+  virtual ~StopwatchSimTime();
+
+public:
+  void SetClock(rclcpp::Clock::SharedPtr _clock);
+
+  /// \brief Start the stopwatch.
+  /// \param[in] _reset If true the stopwatch is reset first.
+  /// \return True if the the stopwatch was started. This will return
+  /// false if the stopwatch was already running.
+
+public:
+  bool Start(const bool _reset = false);
+
+  /// \brief Get the time when the stopwatch was started.
+  /// \return The time when stopwatch was started, or
+  /// std::chrono::steady_clock::time_point::min() if the stopwatch
+  /// has not been started.
+
+public:
+  rclcpp::Time StartTime() const;
```

Comment:
Can this be returned as a const ref not to force a copy?

```suggestion
  const rclcpp::Time & StartTime() const;
```

The same for other getters below

Reply:
Yeah, I think so -- good optimization. I didn't look too closely at the Stopwatch API when I copied it from ignition-math.

Reply:
Ahh yeah it looks like the elapsed times may return locally scoped variables, so we'll need to live with a copy there